### ISSUE TITLE
Dry run fails (false negative) when tx success depends on deployer address

### ISSUE THREAD

**Issue by MichalZalecki** [trufflesuite/truffle#1698]
## Issue

When transaction within the migration depends on deployer's address (e.g. ownership transfer) then dry run fails.

## Steps to Reproduce

Try to transfer ownership as part of the migration.

## Expected Behavior

The same set of `accounts` is used in the dry run as well as during the actual migration.

## Environment

* Operating System: macOS 10.14.3
* Ethereum client: Geth 1.8.22-stable
* Truffle version: v5.0.2
* node version: 10.15.0
* npm version: 6.7.0

**Comment by mrwillis** [comment]
Similar issue here:

https://github.com/trufflesuite/truffle/issues/1612#issuecomment-459273927

**Comment by mrwillis** [comment]
I notice also the use of network names as `network-fork` can really mess up the logic if you have network-dependent logic.

**Comment by mrwillis** [comment]
The best solution I have for anyone facing this issue (which only seems to occur past v5.x truffle) is to just split up your migration files for mainnet with 1-2 transactions each. This way you won't have to restart if something goes wrong.

**Comment by zulhfreelancer** [comment]
Add `skipDryRun: true` in your network config. That should help.

Example:

```
development: {
      skipDryRun: true,
      host: "127.0.0.1",
      port: 7545,
      network_id: "*" // match any network
 },
```

**Comment by MichalZalecki** [comment]
@zulhfreelancer Sure, it's easy to "skip" the problem, but it doesn't solve the problem with dry run.

**Comment by eggplantzzz** [comment]
Hey @MichalZalecki, do you have a specific example that I can take a look at to reproduce the problem?

**Comment by AleG94** [comment]
+1

To reproduce the issue, try to deploy a simple erc20 and call the transfer function from deployer address.
Skipping dry run works.

**Comment by eggplantzzz** [comment]
@seesemichaelj Did you do something related to this issue?

**Comment by seesemichaelj** [comment]
Nope, it hit the back burner

**Comment by eggplantzzz** [comment]
@seesemichaelj Do you have more information about this issue that you could link here?

**Comment by seesemichaelj** [comment]
This is what @nicholasjpaterno wrote on our other repo about the issue:

To add more background info, echoing @seesemichaelj's comments on this issue for solidarity:

web3.eth.getAccounts() with an infura provider returns an empty array, causing https://github.com/trufflesuite/truffle/blob/reporter/packages/truffle-environment/environment.js#L46 to send an empty array to ganache as what accounts are unlocked, causing https://github.com/trufflesuite/ganache-core/blob/283cbeca1cf9f9c598845a00baaa331afc621596/lib/statemanager.js#L327 to be triggered and errors

Proposed Solution:

packages/truffle-environment/environment.js:46
<https://github.com/trufflesuite/truffle|trufflesuite/truffle>trufflesuite/truffle | Added by GitHub
thoughts on making it something more like
```javascript
const options = {
  fork: config.provider,
  gasLimit: block.gasLimit
};

if (accounts.length > 0) {
  options.unlocked_accounts = accounts;
}

config.networks[forkedNetwork] = {
  network_id: config.network_id,
  provider: Ganache.provider(options),
  from: config.from,
  gas: upstreamConfig.gas,
  gasPrice: upstreamConfig.gasPrice
};
```

**Current Status**: Truffle team approves of the changes

**Comment by eggplantzzz** [comment]
So currently Ganache cannot handle an empty array for that option and that is the root of the problem?

**Comment by gnidan** [comment]
@davidmurdoch @nicholasjpaterno we can fix this inside Truffle as proposed above, but y'all really should make Ganache accept an empty array :)

**Comment by davidmurdoch** [comment]
@gnidan, are you suggesting `unlocked_accounts: []` should unlock _all_ accounts instead of no accounts?

**Comment by seesemichaelj** [comment]
This is a discussion that should be on the https://github.com/trufflesuite/ganache-core repo, but the UX would indicate that `unlocked_accounts: []` should unlock no accounts rather than all accounts, which I would argue that the proposed change in Truffle would need to happen as that's not the desired outcome of this scenario

**Comment by eggplantzzz** [comment]
Closing this as I believe it is fixed! Get back to me if you think this is still an issue and we'll re-open. Thanks!
